core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -492,13 +495,152 @@ public void failTaskRemotely(TaskId taskId, Throwable failureCause)
+            }
+        };
+
+        private final String queryId;
+        private final int eventsPerBucket;
+        private long eventsCounter;
+        private long filteredEventsCounter;
+
+        private volatile SoftReference<ListMultimap<String, String>> eventsDebugInfosReference;
+
+        private EventDebugInfos(String queryId, int eventsPerBucket)
+        {
+            this.queryId = requireNonNull(queryId, "queryId is null");
+            this.eventsPerBucket = eventsPerBucket;
+            eventsDebugInfosReference = new SoftReference<>(LinkedListMultimap.create());
+        }
+
+        private boolean add(Event event)
+        {
+            ListMultimap<String, String> eventsDebugInfos = getEventsDebugInfos();
+            String bucket = getBucket(event);
+            Optional<String> debugInfo = getFullDebugInfo(eventsCounter, event);
+            eventsCounter++;
+            if (debugInfo.isEmpty()) {
+                filteredEventsCounter++;
+                return false;
+            }
+            eventsDebugInfos.put(bucket, debugInfo.get());
+            List<String> bucketDebugInfos = eventsDebugInfos.get(bucket);
+            if (bucketDebugInfos.size() > eventsPerBucket) {
+                Iterator<String> iterator = bucketDebugInfos.iterator();
+                iterator.next();
+                iterator.remove();
+            }
+            return true;
+        }
+
+        private ListMultimap<String, String> getEventsDebugInfos()
+        {
+            ListMultimap<String, String> eventsDebugInfos = eventsDebugInfosReference.get();
+            if (eventsDebugInfos == null) {
+                log.info("eventsDebugInfos for %s has been cleared", queryId);
+                eventsDebugInfos = LinkedListMultimap.create();
+                eventsDebugInfosReference = new SoftReference<>(eventsDebugInfos);
+            }
+            return eventsDebugInfos;
+        }
+
+        private String getBucket(Event event)
+        {
+            if (event == Event.WAKE_UP || event == Event.ABORT) {
+                return GLOBAL_EVENTS_BUCKET;
+            }
+            return event.accept(GET_BUCKET_LISTENER);
+        }
+
+        private Optional<String> getFullDebugInfo(long eventId, Event event)
+        {
+            return getEventDebugInfo(event).map(info -> "[" + eventId + "/" + System.currentTimeMillis() + "/" + info + "]");
+        }
+
+        private static Optional<String> getEventDebugInfo(Event event)
+        {
+            if (event == Event.WAKE_UP) {
+                return Optional.of("WAKE_UP");
+            }
+            if (event == Event.ABORT) {
+                return Optional.of("ABORT");
+            }
+            if (event instanceof SplitAssignmentEvent splitAssignmentEvent) {
+                if (splitAssignmentEvent.getAssignmentResult().isEmpty()) {
+                    // there may be significant amount of empty AssignmentResults so lets skip processing of those
```

Comment:
Why do we need to distinguish empty split assignments?
maybe this can be viewed as unnecessary complexity

what if we just recorded all events?
then we wouldn't need those boolean return values

Reply:
Two reasons.
* It could be that scheduler look is not really stuck per se. But we are getting empty events all the time - and it just looks like stuck. We need to notice that, and still log debug information.
* Empty events can push important events out of recorded debug information - making debug logs less useful.

I will add this as a comment